circuit/src/dynamic_sets/mod.rs

```diff
@@ -0,0 +1,912 @@
+					let is_pk_i_y = equal_chip.synthesize(
+						&config.common,
+						&config.main,
+						layouter.namespace(|| "op_pk_j_y == pk_i_y"),
+					)?;
+					let and_chip = AndChipset::new(is_pk_i_x, is_pk_i_y);
+					let is_pk_i = and_chip.synthesize(
+						&config.common,
+						&config.main,
+						layouter.namespace(|| "op_pk_j == pk_i"),
+					)?;
+					let sub_chip = SubChipset::new(one.clone(), is_pk_i);
+					let is_diff_pk = sub_chip.synthesize(
+						&config.common,
+						&config.main,
+						layouter.namespace(|| "op_pk_j != pk_i"),
+					)?;
+
+					// Condition 2. op_pk_j != PublicKey::default()
+					let pk_x_equal_chip =
+						IsEqualChipset::new(pk_x[j].clone(), default_pk_x.clone());
+					let is_default_pk_x = pk_x_equal_chip.synthesize(
+						&config.common,
+						&config.main,
+						layouter.namespace(|| "pk_j_x == default_pk_x"),
+					)?;
+
+					let pk_y_equal_chip =
+						IsEqualChipset::new(pk_y[j].clone(), default_pk_y.clone());
+					let is_default_pk_y = pk_y_equal_chip.synthesize(
+						&config.common,
+						&config.main,
+						layouter.namespace(|| "pk_j_y == default_pk_y"),
+					)?;
+					let and_chip = AndChipset::new(is_default_pk_x, is_default_pk_y);
+					let is_null = and_chip.synthesize(
+						&config.common,
+						&config.main,
+						layouter.namespace(|| "pk_j == default"),
+					)?;
+					let sub_chip = SubChipset::new(one.clone(), is_null);
+					let is_not_null = sub_chip.synthesize(
+						&config.common,
+						&config.main,
+						layouter.namespace(|| "pk_j != default"),
+					)?;
+
+					// Conditions for distributing the score
+					// 1. pk_j != pk_i
+					// 2. pk_j != PublicKey::default()
+					// 3. op_score_sum == 0
+					let and_chip = AndChipset::new(is_diff_pk, is_not_null);
+					let cond = and_chip.synthesize(
+						&config.common,
+						&config.main,
+						layouter.namespace(|| "is_diff_pk && is_not_null"),
+					)?;
+					let and_chip = AndChipset::new(cond, is_sum_zero.clone());
+					let cond = and_chip.synthesize(
+						&config.common,
+						&config.main,
+						layouter.namespace(|| "is_diff_pk && is_not_null && is_sum_zero"),
+					)?;
+					let select_chip = SelectChipset::new(cond, one.clone(), ops_i[j].clone());
+					ops_i[j] = select_chip.synthesize(
+						&config.common,
+						&config.main,
+						layouter.namespace(|| "filtered op score"),
+					)?;
+				}
+
+				// Add to "filtered_ops"
+				filtered_ops.push(ops_i);
+			}
+
+			filtered_ops
+		};
+
+		// "Normalization"
+		let ops = {
+			let mut normalized_ops = Vec::new();
+			for i in 0..NUM_NEIGHBOURS {
+				let mut ops_i = Vec::new();
+
+				// Compute the sum of scores
+				let mut op_score_sum = zero.clone();
+				for j in 0..NUM_NEIGHBOURS {
+					let add_chip = AddChipset::new(op_score_sum.clone(), ops[i][j].clone());
+					op_score_sum = add_chip.synthesize(
+						&config.common,
+						&config.main,
+						layouter.namespace(|| "op_score_sum"),
+					)?;
+				}
```

Comment:
The sum is already calculated once here: https://github.com/eigen-trust/protocol/pull/197/files#diff-d1a67c0b28a8f922a0aafb9671a55106124235c920ec65ab43c1ea9f883b67b1R494-R502

Perhaps we could reuse it

Reply:
Hmm, I don't think we can reuse it.
The `op_score_sum` could change since there is logic of assigning `Fr::one()` to opinion. (https://github.com/eigen-trust/protocol/blob/dynamic-sets-circuit/circuit/src/dynamic_sets/mod.rs#L569-L590).
Hence, I think we should re-compute the `op_score_sum`.

Reply:
Right, sorry I overlooked that.